examples/2.0.0/aws-sns-publish.yml

```diff
@@ -0,0 +1,46 @@
+asyncapi: 2.0.0
+id: 'https://github.com/asyncapi/asyncapi/blob/master/examples/2.0.0/aws-sns-publish.yml'
+info:
+  title: Simple AWS SNS Publish specification
+  description: |
+    Publishes a hello-world message to a particular queue
+  version: '0.1.0'
+servers:
+  test:
+    # This URL doesn't really describe a test server, but it does describe the
+    # top-level URL used to post a message to a given AWS region
+    url: https://sns.us-east-1.amazonaws.com/
+    description: Test environment for message publishing
+    protocol: sns
+    protocolVersion: "2010-03-31" #Can we reuse the "version" from down below?
+    bindings:
+      # Details on topicARNs: https://docs.aws.amazon.com/sns/latest/dg/sns-tutorial-create-topic.html
+      # This is a prefix, not a complete topicARN, but the prefix identifies
+      # the "server" environment (which is really just the account where this topic lives)
+      topicArnPrefix: "arn:aws:sns:us-east-1:123456789012:"
+channels:
+  helloWorld:
+    description: Topic used for hello world messages
+    bindings:
+      sns:
+        # This should be appended to the servers' topicARNPrefix to get the complete topicARN
+        # Can we do this automatically using JSON-Schema trickery somehow?
+        topicArnTopicName: "helloWorld"
+    publish:
+      message:
+        # The payload is equivalent to the AWS "Message" field
+        # Publish documentation: https://docs.aws.amazon.com/sns/latest/api/API_Publish.html
+        payload:
+          type: string
+          pattern: '^hello .+$'
+      bindings:
+        sns:
+          Action: Publish
+          # This is a concatenation of topicArnPrefix + topicArnTopicName from servers/channels above
+          TopicArn: "arn:aws:sns:us-east-1:123456789012:helloWorld"
```

Comment:
I think the whole `TopicArn` is redundant, since the template/client should use the `topicArnPrefix` together with this  `TopicArn` variable. Therefore I suggest removing `arn:aws:sns:us-east-1:123456789012:` to it only contain `helloWorld`.

Reply:
Actually, I don't think that `helloWorld` should even be there since that would always be the name of the channel. Am I right?

IMO, all this data from the topicArn should be included as an AWS extension and not as part of the protocol binding. This data has nothing to do with the protocol but more with infrastructure/user details.

What do you think?

Reply:
As this is the SNS protocol, it seems like part of the bindings (SNS is AWS-only) - so everything that is AWS specific *should* go IMO into the SNS binding, no?

As for the 'name' of the channel, that's another question I have - because the name of the channel can be sometimes different across different environments, even though these channels have the same semantic meaning - e.g. maybe the channel is the 'helloWorld' channel but the actual topic in SNS is called 'helloWorld-dev-1818727' and 'helloWorld-test-11288273' or something.

Reply:
> because the name of the channel can be sometimes different across different environments, even though these channels have the same semantic meaning

You just said it yourself. The topic name includes the environment, which is not part of the protocol but part of their product. By "protocol" I mean the "wire protocol", and bindings should only have the protocol features, not product features. Think about this way: tomorrow AWS starts offering to connect to their services using AMQP instead of their proprietary protocol. You still have environments but environments are not a feature of the AMQP protocol. Sure the information about the environment will be carried somewhere. The difficulty with AWS is that they have everything (infrastructure, product, and protocol) mixed up and bundled as a single thing.

Therefore, your channel or topic is actually `helloWorld`. If you deploy the same app to another region, environment or using another user, your channel should be the same. If that's not true, then maybe you'd need to have a different AsyncAPI file per environment and user, which I don't think is what you want.

For instance, given the following file:

```yaml
...
servers:
  dev:
    url: yoursnsurl.com
    protocol: sns
    x-aws-sns:
      environment: dev
      user: 1818727
  production:
    url: yoursnsurlprod.com
    protocol: sns
    x-aws-sns:
      environment: prod
      user: 1818727
channels:
  helloWorld:
    bindings:
      sns:
        Action: Publish
        # ...and more stuff here
```

The code generator and docs generator can process the information and give you the topicArn as `helloWorld-dev-1818727` or `helloWorld-prod-1818727` (depending on the server you choose), which is what you'll use in your code.

Reply:
Ah ok, I'm sorry my example above was probably confusing.

AWS itself does not have any restrictions on the topic name (see https://aws.amazon.com/sns/faqs/#Features_and_functionality)

I've pushed another commit that simplifies the bindings somewhat.

The main thing we need to be able to communicate through the spec is the topicArn for a given channel, while noting that this topicArn is a sort of 'global' locator for Amazon that uniquely identifies a *particular* channel, which is a combination of "AWS account" + "region" + "topicName".

Reply:
Also another note re: this:
'By "protocol" I mean the "wire protocol", and bindings should only have the protocol features, not product features.'

So maybe a core issue here is that SNS isn't really a 'wire protocol', per se. Rather, it is a product, that you interact with via HTTPS with certain parameters (to send messages to), and it in term can distribute messages over a number of different protocols, including SQS, HTTPS, SMTP, SMS, etc.